Rate limit slow price estimators

MartinquaXD opened the pull request:
Fixes #244 

This PR allows us to rate limit price estimator which are currently taking a very long time to respond.
To do that I made the `RateLimiter` a little more generic such that it is able to rate limit any task (not just http requests).
That is accomplished by not only supplying a future to be executed (`task`) but also a function which looks at the result and determines whether rate limiting is required.
That way the paraswap API can check for status code 429 on HTTP responses and price estimator can check whether requests took longer than a user defined amount of time.

Note that the `RateLimiter` already logs relevant events and records metrics so we can implement alerts based on that immediately.

Misc changes:
- moved name from `RateLimitingStrategy` into `RateLimiter` and updated APIs accordingly.

### Test Plan
Manual test

MartinquaXD commented:
Given the fact that the `RateLimiter` is now quite generic and not strictly related to HTTP it would be reasonable to move it out of the `http_client` file and somewhere else.
👍 keep it; 👎 move it

vkgnosis commented:
I left a [comment ](https://github.com/cowprotocol/services/issues/244#issuecomment-1152132112)on the issue discussing this type of implementation versus a simple timeout. Didn't want to post here because it really isn't about this specific PR and more about the general issue.

nlordell commented:
I also in general agree with change so all price estimates aren't bottle-necked by the slowest one.

nlordell commented on crates/shared/src/price_estimation/rate_limited.rs:
```diff
@@ -0,0 +1,162 @@
+    time::{Duration, Instant},
+};
+
+pub trait RateLimitedPriceEstimatorExt {
```
I really like how expressive this extension trait makes the code:
```
SomeEstimator::new(...).rate_limited(...)
```

So much style 😎

MartinquaXD replied:
Glad you like it.
I was considering applying this pattern for all the price estimators to see if it would make creating the price estimator hierarchy in the [order book](https://github.com/cowprotocol/services/blob/main/crates/orderbook/src/main.rs#L528-L636) look more reasonable.

vkgnosis replied:
I prefer standalone functions over extension traits usually. Would only make it a trait if it is used so much that the small ergonomic improvement trumps the more complicated construction. Can still keep it this way here.

nlordell reviewed (approved):
Code looks good to me.

MartinquaXD commented:
- moved `RateLimiter` into its own file
- replaced `ignored` paraswap test with unit test for `RateLimiter`
- discovered/fixed subtle bug causing the smallest back off to be actually `min_back_off * back_off_growth_factor` instead of `min_back_off`

vkgnosis commented on crates/shared/src/price_estimation/rate_limited.rs:
```diff
@@ -0,0 +1,162 @@
+        &'a self,
+        queries: &'a [Query],
+    ) -> BoxStream<'_, (usize, super::PriceEstimateResult)> {
+        let start = Instant::now();
+        futures::stream::iter(queries.iter())
+            .enumerate()
+            .then(move |(index, query)| async move {
+                let result = self
+                    .rate_limiter
+                    .execute(single_estimate(&*self.inner, query), |_| {
+                        let elapsed = start.elapsed();
+                        elapsed > self.healthy_response_time
+                    })
+                    .await
+                    .map_err(PriceEstimationError::from);
+
+                (index, result)
+            })
```
Another observation in favor of this:
Imagine our healthy time is 10 seconds and we want 10 estimates. Then if we call an estimator one by one it would be allowed to return each of those results after 10 seconds for a total of 100 seconds and be considered healthy. But if we called in a batching way then only the first response would be healthy because for the remaining 9 with the current code they are considered out of time.

MartinquaXD replied:
Those are good points.
I decided to go for the option of rate limiting every estimator except `Baseline`.

vkgnosis replied:
I still think it would be cleaner to integrate with "single price estimators" directly rather than keeping this code. I am also fine with doing that later.

MartinquaXD replied:
I looked into how that could be done and I believe the error propagation of `RateLimiter` needs to be improved in order to be compatible with that idea.
On the other hand the way the error propagation happens right now is not ideal anyway so I might as well fix that and then integrate the `RateLimiter` into each `PriceEstimator`.

MartinquaXD commented:
Merged main => updated argument formatting
Did a manual test run, seems to work as expected.

vkgnosis commented on crates/orderbook/src/arguments.rs:
```diff
@@ -280,6 +289,9 @@ impl std::fmt::Display for Arguments {
             "fast_price_estimation_results_required: {}",
             self.fast_price_estimation_results_required
         )?;
+        write!(f, "price_estimation_rate_limiter: ")?;
+        display_option(&self.price_estimation_rate_limiter, f)?;
+        writeln!(f)?;
```
It is also okay to do 

```rust
writeln!(
            f,
            "price_estimation_rate_limiter: {:?}",
            self.price_estimation_rate_limiter
        )?;
```

If you are fine with the default Debug implementation of the struct and don't want to impl Display for it.

MartinquaXD replied:
I had that initially but that didn't look very nice.
This is how it looks now with only the configured values displayed:
```
...
fast_price_estimation_results_required: 2
price_estimation_rate_limiter: RateLimitingStrategy{ min_back_off: 60s, max_back_off: 60s, growth_factor: 2.0 }
token_detector_fee_values: Static
...
```
